### PR TITLE
fix: Add --probabilities flag to csoaa, Resolves #3331

### LINKED ISSUE #3331: CSOAA in Python not returning probabilities
Hello,

Thanks for the great work on VW over the many years!

I'm using the Python package with the `csoaa` option to deal with a multilabel problem. It seems to train fine, but I can't get it to return probabilities. This is how I trained without any errors:

```
from vowpalwabbit import pyvw

vw = pyvw.vw(quiet=True, csoaa=y_train.shape[1], b=26, ngram="d2", probabilities=True)
passes = 1
for n in range(passes):
    print(n)
    if n >= 1:
        examples = shuffle(Xy_train)
    else:
        examples = Xy_train
    for idx, example in enumerate(examples):
        vw.learn(example)
```

Where `Xy_train` has things like 

```
Xy_train[0]
Out: '1:1.0 2:1.0 3:1.0 4:0.0 5:1.0 6:1.0 7:1.0 8:1.0 9:1.0 10:1.0 11:1.0 12:1.0 13:1.0 14:1.0 15:1.0 16:1.0 17:1.0 18:1.0 19:1.0 20:1.0 21:1.0 22:1.0 23:1.0 24:1.0 |text a transient based real time scheduling algorithm in fms 
```

At test time, I have data like:

```
Xy_val[0]
Out: '1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 |text covid associated ards treated with dexamethasone codex study design and rationale for a randomized trial
```

And doing `vw.predict(Xy_val[0])` probabilities aren't returned, but just a single integer. How can one get probabilities out of this? I've tried `vw.predict(Xy_val[0], i)` for various `i` and no luck.

Thank you.

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: vowpalwabbit/csoaa.cc [hunk 1]
```diff
@@ -129,7 +145,69 @@ void predict_or_learn(csoaa& c, single_learner& base, example& ec)
...
+    // === Compute multiclass_log_loss
+    // TODO:
+    // What to do if the correct label is unknown, i.e. (uint32_t)-1?
+    //   Suggestion: increase all.sd->weighted_unlabeled_examples???,
+    //               but not sd.example_number, so the average loss is not influenced.
+    // What to do if the correct_class_prob==0?
+    //   Suggestion: have some maximal multiclass_log_loss limit, e.g. 999.
+    float multiclass_log_loss = 999;  // -log(0) = plus infinity
+    float correct_class_prob = 0;
+    if (ec.l.multi.label <= c.num_classes)  // prevent segmentation fault if labeĺ==(uint32_t)-1
```
Comment (jackgerrits) [inline vowpalwabbit/csoaa.cc:162]: Is uint32_t a special value?

File: vowpalwabbit/csoaa.cc [hunk 2]
```diff
@@ -129,7 +145,69 @@ void predict_or_learn(csoaa& c, single_learner& base, example& ec)
...
+    prediction++;  // prediction is 1-based index (not 0-based)
+    float zero_one_loss = 0;
+    if (ec.l.multi.label != prediction) zero_one_loss = ec.weight;
+
+    // === Print probabilities for all classes
+    std::ostringstream outputStringStream;
+    for (uint32_t i = 0; i < c.num_classes; i++)
+    {
+      if (i > 0) outputStringStream << ' ';
+      if (all.sd->ldict) { outputStringStream << all.sd->ldict->get(i + 1); }
```
Comment (jackgerrits) [inline vowpalwabbit/csoaa.cc:185]: I don't think ldict is used in csoaa

File: vowpalwabbit/csoaa.cc [hunk 3]
```diff
@@ -129,7 +145,69 @@ void predict_or_learn(csoaa& c, single_learner& base, example& ec)
...
 }
 
-void finish_example(vw& all, csoaa&, example& ec) { COST_SENSITIVE::finish_example(all, ec); }
+template <bool probabilities>
+void finish_example(vw& all, csoaa& c, example& ec)
+{
+  if (probabilities)
+  {
+    // === Compute multiclass_log_loss
+    // TODO:
```
Comment (jackgerrits) [inline vowpalwabbit/csoaa.cc:154]: Would be nice if we could either resolve the todos or determine if they are relevant before copying them over here

File: vowpalwabbit/csoaa.cc [hunk 4]
```diff
@@ -77,11 +80,24 @@ void predict_or_learn(csoaa& c, single_learner& base, example& ec)
...
 
   bool dont_learn = DO_MULTIPREDICT && !is_learn;
+  if (probabilities)
+  {
+    base.multipredict(ec, 0, c.num_classes, c.pred, true);
+    for (uint32_t i = 0; i < c.num_classes; i++) ec.pred.scalars.push_back(c.pred[i].scalar);
+    float sum_prob = 0;
+    for (uint32_t i = 0; i < c.num_classes; i++)
+    {
+      ec.pred.scalars[i] = 1.f / (1.f + correctedExp(-c.pred[i].scalar));
```
Comment (jackgerrits) [inline vowpalwabbit/csoaa.cc:90]: Would be nice to add a `normalize` function to [`VW::math`](https://vowpalwabbit.org/docs/vowpal_wabbit/cpp/latest/namespaceVW_1_1math.html) since this seems to be done a few times over the codebase

File: test/train-sets/ref/help.stdout [hunk 5]
```diff
@@ -429,6 +429,7 @@ Cost-sensitive Active Learning:
   --csa_debug                           print debug stuff for cs_active
 Cost Sensitive One Against All:
   --csoaa arg           One-against-all multiclass with <k> costs
+  --probabilities       predict probabilites of all classes
```
Comment (olgavrou) [inline test/train-sets/ref/help.stdout:432]: does this predict probabilities or report probabilities?

Comment (bassmang) [reply test/train-sets/ref/help.stdout:432]: I believe it's doing doing both, but this is following the naming convention for oaa and csoaa_ldf.

File: vowpalwabbit/csoaa.cc (hunk 4 above)
Comment (bassmang) [reply vowpalwabbit/csoaa.cc:90]: Yea I agree, I'll add that in

File: vowpalwabbit/csoaa.cc (hunk 1 above)
Comment (bassmang) [reply vowpalwabbit/csoaa.cc:162]: There does seem to be a mismatch, I can take this comment out

File: vowpalwabbit/csoaa.cc (hunk 2 above)
Comment (bassmang) [reply vowpalwabbit/csoaa.cc:185]: This is used for the named case, ie in `test/pred-sets/ref/csoaa_named_probabilities.predict` it will output the names of the labels instead of the indices
